numba/core/interpreter.py

```diff
@@ -314,6 +314,164 @@ def peep_hole_delete_with_exit(func_ir):
+                func_name = stmt.value.func.name
+                # If we have an update or a setitem
+                # it will be the previous expression.
+                getattr_stmt = blk.body[i - 1]
+                args = stmt.value.args
+                if (
+                    isinstance(getattr_stmt, ir.Assign)
+                    and getattr_stmt.target.name == func_name
+                    and isinstance(getattr_stmt.value, ir.Expr)
+                    and getattr_stmt.value.op == "getattr"
+                    and getattr_stmt.value.value.name in lit_maps
+                ):
+                    update_map_name = getattr_stmt.value.value.name
+                    attr = getattr_stmt.value.attr
+                    if attr == "__setitem__":
+                        append_inst = False
+                        # If we have a setitem, update the lists
+                        map_updates[update_map_name].append(args)
+                        # Remove the setitem
+                        new_body.pop()
+
+                    elif attr == "update" and args[0].name in lit_maps:
+                        append_inst = False
+                        # If we have an update and the arg is also
+                        # a literal dictionary, fuse the lists.
+                        map_updates[update_map_name].extend(
+                            map_updates[args[0].name]
+                        )
+                        # Remove the update
+                        new_body.pop()
+
+            if append_inst:
+                # Check if we need to pop any dictionaries from the const map.
+                # Skip the setitem and update that will be used later.
+                if not (
+                    isinstance(stmt, ir.Assign)
+                    and isinstance(stmt.value, ir.Expr)
+                    and stmt.value.op == "getattr"
+                    and stmt.value.value.name in lit_maps
+                    and stmt.value.attr in ("__setitem__", "update")
+                ):
+                    for var in stmt.list_vars():
+                        # If is used pop it from the const map
+                        if var.name in lit_maps:
+                            lineno = lit_maps[var.name]
+                            items = map_updates[var.name]
+                            new_body.append(
+                                build_new_buildmap(var.name, blk, lineno, items)
+                            )
+                            # Delete from the const maps
+                            del lit_maps[var.name]
+                            del map_updates[var.name]
+                new_body.append(stmt)
+        # Update we any dictionaries that weren't used in the block yet.
+        # We pop off the last statement because this needs to remain last
+        # as it might be a jump, branch, or return
+        end = new_body.pop()
+        for var_name, lineno in lit_maps.items():
+            items = map_updates[var_name]
```

Comment:
I think I may need to also update definitions whenever a node is deleted or when adding the new build maps (since the build_map changes to include more information).

Reply:
Not completely sure, but it seems like the map definitions should be updated based on the other optimizations.

Reply:
Assuming this is referring to `func_ir._definitions`, then I think the definitions need to be kept consistent. I don't think it hugely matters whether this is fixed at the time of invalidation or buffered and fixed at the end of the transform, as nothing else will be accessing it until the transform is complete. The `list_to_tuple` rewrite elects to do this at invalidation time, e.g. (and others): https://github.com/numba/numba/blob/f04f4b164887e5aa5ffeeed1bef317e13ec995f8/numba/core/interpreter.py#L176-L182